### DISCUSSION TITLE
$nextTick fires before rendering completes with nested x-data scopes in V3

### DISCUSSION THREAD

**Discussion by atomgiant** [alpinejs/alpine#1806] [7. Bugs (archived)]
$nextTick in Alpine V3 is firing before the rendering changes are complete. I believe this is only happening if the $nextTick is inside a nested x-data scope.

I put together this pen to demonstrate this. The select menu uses $nextTick to reposition the menu and focus the selected element when the menu opens:

https://codepen.io/atomgiant/pen/rNmypee

You may have to click the selected item in the menu several times to see the $nextTick fire before the DOM fully renders.

This code did not have this issue in V2. Also, if you remove the parent x-data this issue does not happen. I suspect maybe the shared scope is causing $nextTick to fire when the parent updates before the child $nextTick completes.

**Comment by KevinBatdorf** [comment]
Any chance you can simplify the demo and make the symptoms more obvious with something like console logs?

**Reply by atomgiant** [reply]
Hi @KevinBatdorf - thanks for taking a look at this.

I just updated the code to log info to the console to make it easier to see what is happening.

You may need to click the menu item several times to see this happen. Here is an example of the console.log when this is happening. 
 
![Screen Shot 2021-07-17 at 4 08 55 PM](https://user-images.githubusercontent.com/21659/126048292-4f0db202-e4f6-4b33-8d2f-04da3eac78d3.png)

The $refs.menu is not visible yet when the $nextTick runs even though the x-show="open" variable is true. You can also see this in the "display:none" style attribute for the $refs.menu element.

**Reply by KevinBatdorf** [reply]
I can see it now. Any reason you're using `offsetParent` there? It will probably work if you just use `parentElement` or not even check that at all? Even when I see the error logs, the item is still in focus.

Accessing the `offsetParent` property causes browser page reflow so that's probably the issue.

**Reply by atomgiant** [reply]
offsetParent is just how I am checking if the item is visible or not. 

The issue is $nextTick is calling my function before the this.$refs.menu is visible when it should be from the x-show for that same tick. This makes it so the this.$refs.menu.getBoundingClientRect() doesn't work since the item is not visible. 

It is subtle but if you look at the menu when this happens the focus is not happening. The menu does not shift to the selected item, and the selected item does not have focus (which gives it a gray background).

So in summary, this.$refs.menu has an x-show set to show if `open` is true. $nextTick is happening before the x-show is finished when this race condition happens.

**Reply by KevinBatdorf** [reply]
So the way it works in JS as I understand things, is JS specific code is run line by line, and every web API call is pushed to the event stack to run later. `$nextTick` is just pushing the callback you pass in to the end of that stack, off the main thread. 

If you have a lot of other events being pushed to the event stack that cause the browser repaint/reflow, it could delay the visibility of the node, and thus cause `offsetParent` (which causes its own reflow) to report null.

If the issue is Alpine specific, you should be able to reproduce it in a much smaller example. If you can't, then I'd focus on just optimizing your code to eliminate reaching into the dom and using any web APIs. I haven't investigated too much ,to be fair, so there could be something, but that's why a minimal example helps.

I could be wrong entirely, but that's my current thinking. A more minimal foo/bar example will help though

**Reply by atomgiant** [reply]
Thanks @KevinBatdorf. I created this simpler example https://codepen.io/atomgiant/pen/KKmqGpw. Unfortunately it almost never fails on Codepen, but fails roughly 5 percent of the time running locally. 

I believe there may be some kind of issue here related to the timing of x-show and $nextTick. This is tricky to reproduce in a simple way reliably.

The good news is I found a reasonable workaround. Instead of using x-show with $nextTick I am rewriting my JavaScript to explicitly show / hide the menu by adding / removing CSS classes. That way I can guarantee the menu is visible when focusing the content.
